src/client/datascience/kernel-launcher/kernelLauncher.ts

```diff
@@ -84,6 +92,42 @@ export class KernelLauncher implements IKernelLauncher {
 
     @captureTelemetry(Telemetry.KernelLauncherPerf)
     public async launch(
+        kernelConnectionMetadata: KernelSpecConnectionMetadata | PythonKernelConnectionMetadata,
+        timeout: number,
+        resource: Resource,
+        workingDirectory: string,
+        cancelToken?: CancellationToken
+    ): Promise<IKernelProcess> {
+        // If this is a python interpreter, make sure it has ipykernel
+        if (kernelConnectionMetadata.interpreter) {
+            await this.installDependenciesIntoInterpreter(kernelConnectionMetadata.interpreter, cancelToken);
+        }
+
+        // Should be available now, wait with a timeout
+        const result = await waitForPromise(
+            Promise.race([
+                this.launchProcess(kernelConnectionMetadata, resource, workingDirectory),
+                createPromiseFromCancellation({
+                    cancelAction: 'reject',
+                    defaultValue: new CancellationError(),
+                    token: cancelToken
+                })
+            ]),
+            timeout
+        );
+
+        if (result instanceof CancellationError) {
+            throw result;
+        }
+
+        if (!result) {
+            throw new TimedOutError(localize.DataScience.rawKernelProcessNotStarted());
```

Comment:
This is a timeout error or `KernelLaunchError`. There's no timeout component here, its just that the kernel didn't launch & we don't know why. Isn't it.

Reply:
Please ignore the above.

Reply:
No if the result is undefined it's a timeout error (waitForPromise returns 'null' on timeout). Kernel launch failure will throw an exception. this.launchProcess always returns an object.

Reply:
I'll add a comment to that effect

Reply:
I never did like the waitForPromise. ideally i'd like that to throw an error instead of always checking the return value. This way its easier, else everyhere this is used we have to do the same thing... I think ther'es an issue open to fix how waitForPromise works.

Reply:
Anyways, not nothing that needs to be fixed now.

Reply:
Could make it explicit with === null, not just a falsy check. I see that most of the other locations that we use waitForPromise (jupyterNotebook, rawJupyterSession) we do an explicit null check for the timeout case. It works now with the falsy check, but if something was refactored to possibly return undefined then it would go down the wrong path here.

Reply:
This isn't being used anymore so don't have to worry about it. I agree we should change waitForPromise to reject on timeout instead of returning null